src/utils/ObjectUtils.ts

```diff
@@ -63,6 +63,27 @@ export function shallowEqual(obj1: any, obj2: any): Boolean {
   if (obj1 === obj2) {
     return true;
   }
+
+  // Special handling for Immutables, as they must be handled specially.
+  // While this technically means this is deep equality for Immutables,
+  // it's better to have a more specific check than an entirely incorrect one.
+  if (
+    typeof obj1.hashCode === 'function' &&
+    typeof obj2.hashCode === 'function'
+  ) {
+    // `hashCode` is guaranteed to be the same if the objects are the same, but
+    // is NOT guaranteed to be different if the objects are different (hash
+    // collisions are possible). If the hash codes are different, then we can
+    // preemptively return false as a performance optimization. Otherwise,
+    // return the result of a call to `equals`.
+    // https://github.com/immutable-js/immutable-js/blob/59c291a2b37693198a0950637c5d55cd14dd6bc4/src/is.js#L52-L55
+    if (obj1.hashCode() !== obj2.hashCode()) {
+      return false;
+    } else {
+      return obj1.equals(obj2);
```

Comment:
Should check that equals is a func as well.

Reply:
according to the documentation in immutable, if an object implements `hashCode` it must also implement `equals`. do we want to supoort objects that don't conform to that?

Reply:
I mean there could be some other random object with hash codes but not equals. It would be odd but the check is basically free.